src/d3a/models/market/market_structures.py

```diff
@@ -263,46 +255,56 @@ class Trade:
     seller_id: str = None
     buyer_id: str = None
 
-    def __str__(self):
+    def __str__(self) -> str:
         return (
             "{{{s.id!s:.6s}}} [origin: {s.seller_origin} -> {s.buyer_origin}] "
-            "[{s.seller} -> {s.buyer}] {s.offer.energy} kWh @ {s.offer.price} {rate} "
-            "{s.offer.id} [fee: {s.fee_price} cts.]".
-            format(s=self, rate=round(self.offer.energy_rate, 8))
+            "[{s.seller} -> {s.buyer}] {s.offer_bid.energy} kWh @ {s.offer_bid.price} {rate} "
+            "{s.offer_bid.id} [fee: {s.fee_price} cts.]".
+            format(s=self, rate=round(self.offer_bid.energy_rate, 8))
         )
 
     @classmethod
-    def _csv_fields(cls):
+    def _csv_fields(cls) -> Tuple:
         return (tuple(cls.__dataclass_fields__.keys())[1:2] + ("rate [ct./kWh]", "energy [kWh]") +
                 tuple(cls.__dataclass_fields__.keys())[3:5])
 
-    def _to_csv(self):
-        rate = round(self.offer.energy_rate, 4)
+    def _to_csv(self) -> Tuple:
```

Comment:
Nit: two points about this method:
1. the method name is a bit misleading. The output is not a CSV file, but rather a tuple that can be used for many purposes (one of those is to create a CSV file);
2. the method should be public, since it's being used outside of this class

I know this is not part of the PR, but I'm taking this opportunity that you gave us with the refactoring to note it.